### PR TITLE
Make vectorization of calculating checksum between replicas

### PR DESCRIPTION
**Pull request by chaoyli**
Now, when check consistency, the calculate checksum is non-vectorized.
This pull request is to support the vectorized consistency checksum algorithm.

### CODE REVIEW

File: be/src/storage/task/engine_checksum_task.cpp [hunk 1]
```diff
@@ -64,69 +63,50 @@ OLAPStatus EngineChecksumTask::_compute_checksum() {
...
-    while (true) {
-        OLAPStatus res = reader.next_row_with_aggregation(&row, mem_pool.get(), agg_object_pool.get(), &eof);
-        if (res == OLAP_SUCCESS && eof) {
-            VLOG(3) << "reader reads to the end.";
-            break;
-        } else if (res != OLAP_SUCCESS) {
-            LOG(WARNING) << "fail to read in reader. res=" << res;
-            return res;
+    auto chunk = vectorized::ChunkHelper::new_chunk(child_schema, reader_params.chunk_size);
+    Status status = reader->get_next(chunk.get());
```
Comment (sduzh) [inline be/src/storage/task/engine_checksum_task.cpp:103]:
```suggestion
    st = reader->get_next(chunk.get());
```

File: be/src/storage/task/engine_checksum_task.cpp [hunk 2]
```diff
@@ -64,69 +63,50 @@ OLAPStatus EngineChecksumTask::_compute_checksum() {
...
     uint32_t row_checksum = 0;
-    while (true) {
-        OLAPStatus res = reader.next_row_with_aggregation(&row, mem_pool.get(), agg_object_pool.get(), &eof);
-        if (res == OLAP_SUCCESS && eof) {
-            VLOG(3) << "reader reads to the end.";
-            break;
-        } else if (res != OLAP_SUCCESS) {
-            LOG(WARNING) << "fail to read in reader. res=" << res;
-            return res;
+    auto chunk = vectorized::ChunkHelper::new_chunk(child_schema, reader_params.chunk_size);
```
Comment (sduzh) [inline be/src/storage/task/engine_checksum_task.cpp:102]: `child_schema` has been `std::move()`ed, you should not use it.

File: be/src/storage/task/engine_checksum_task.cpp [hunk 3]
```diff
@@ -64,69 +63,50 @@ OLAPStatus EngineChecksumTask::_compute_checksum() {
...
-    }
+    vectorized::Schema child_schema =
+            vectorized::ChunkHelper::convert_schema_to_format_v2(tablet_schema, return_columns);
 
-    res = reader.init(reader_params);
-    if (res != OLAP_SUCCESS) {
-        LOG(WARNING) << "initiate reader fail. res=" << res;
-        return res;
+    std::shared_ptr<vectorized::Reader> reader =
+            std::make_shared<vectorized::Reader>(tablet, Version(0, _version), std::move(child_schema));
```
Comment (sduzh) [inline be/src/storage/task/engine_checksum_task.cpp:84]: `vectorized::Reader` has been renamed to `vectorized::TabletReader` in the main branch.

**Review by sduzh — DISMISSED** [review]

File: be/src/storage/task/engine_checksum_task.cpp [hunk 4]
```diff
@@ -64,73 +63,70 @@ OLAPStatus EngineChecksumTask::_compute_checksum() {
...
-            VLOG(3) << "reader reads to the end.";
-            break;
-        } else if (res != OLAP_SUCCESS) {
-            LOG(WARNING) << "fail to read in reader. res=" << res;
-            return res;
+
+    uint32_t checksum = 0;
+    uint32_t num_rows = 0;
+    uint32_t hash_codes[config::vector_chunk_size];
+    memset(hash_codes, 0, sizeof(uint32_t) * config::vector_chunk_size);
```
Comment (DeepThinker666) [inline be/src/storage/task/engine_checksum_task.cpp:106]: is it necessary to init the array to 0?

File: be/src/storage/task/engine_checksum_task.cpp [hunk 5]
```diff
@@ -64,73 +63,70 @@ OLAPStatus EngineChecksumTask::_compute_checksum() {
...
-        // the memory allocate by mem pool has been copied,
-        // so we should release memory immediately
-        mem_pool->clear();
-        agg_object_pool = std::make_unique<ObjectPool>();
+        chunk->reset();
+        st = reader.get_next(chunk.get());
+    }
+
+    if (!st.is_end_of_file() && !st.ok()) {
+        LOG(INFO) << "Failed to do checksum. error:=" << st.to_string();
```
Comment (DeepThinker666) [inline be/src/storage/task/engine_checksum_task.cpp:124]: LOG(WARNING)

Comment (DeepThinker666) [reply be/src/storage/task/engine_checksum_task.cpp:124]: better to print tablet info for debug

**Review by sduzh — APPROVED** [review]

**Review by imay — APPROVED** [review]

**Pull request merged**